lib/src/generator_database.dart

```diff
@@ -138,4 +140,119 @@ class GeneratorDatabase {
+      return _allPropsCache[statement.name]!;
+    }
+
+    final resolvedGenerics = typeDescriptorOf(statement)
+            ?.genericParameters
+            .fold<Map<String, TypeRef>>(
+          {},
+          (acc, curr) => {
+            ...acc,
+            // initially, generics should resolve to a generic
+            curr.name: TypeRef(
+              generic: TypeRef_Generic(name: curr.name),
+              nullable: false,
+            ),
+          },
+        ) ??
+        {};
+
+    return _allPropsCache[statement.name] =
+        _allPropertiesOfAux(statement, resolvedGenerics);
+  }
+
+  TypeRef _resolveType(TypeRef type, Map<String, TypeRef> generics) {
+    if (type.hasGeneric()) {
+      final arg = generics[type.generic.name];
+      if (arg == null) return type;
+
+      return arg
+        ..freeze()
+        ..rebuild((arg) {
+          arg.nullable = type.nullable || arg.nullable;
+        });
+    } else if (type.hasInternal()) {
+      return TypeRef(
+        nullable: type.nullable,
+        internal: TypeRef_Internal(
+          name: type.internal.name,
+          arguments: type.internal.arguments
+              .map((type) => _resolveType(type, generics)),
+        ),
+      );
+    } else if (type.hasKnown()) {
+      return TypeRef(
+        nullable: type.nullable,
+        known: TypeRef_Known(
+          type: type.known.type,
+          arguments:
+              type.known.arguments.map((type) => _resolveType(type, generics)),
+        ),
+      );
+    } else {
+      throw StateError('Unhandled TypeRef type');
+    }
+  }
+
+  /// Given a statement and a map of resolving generics (for example {'T': List<int>})
+  /// returns all properties of the statement with resolved generics
+  List<PropertyRef> _allPropertiesOfAux(
+    Statement statement,
+    Map<String, TypeRef> resolvedGenerics,
+  ) {
+    final typeDescriptor = typeDescriptorOf(statement);
+
+    if (typeDescriptor == null) {
+      return [];
+    }
+
+    final properties = typeDescriptor.extends_1
+        .where((typeRef) => typeRef.hasInternal())
+        .expand((typeRef) {
+          final internal = typeRef.ensureInternal();
+          final statement = find(internal.name);
+          if (statement == null) return <PropertyRef>[];
+
+          final resolvedGenerics = typeDescriptorOf(statement)
+                  ?.genericParameters
+                  .foldIndexed<Map<String, TypeRef>>(
+                {},
+                (i, acc, curr) => {
+                  ...acc,
+                  // get the concrete type from child's generic argument list
+                  curr.name: internal.arguments[i],
+                },
+              ) ??
+              {};
+
+          return _allPropertiesOfAux(statement, resolvedGenerics);
+        })
+        .followedBy(typeDescriptor.properties)
+        // remove duplicate properties which come from implementing an interface
+        .groupFoldBy(
```

Comment:
This should not happen, at least in the design (I can see when it might happen on C# side, but we should disallow that there, not overcome it here). Did you stumble on it in some case or just wanted to be sure?

Reply:
Followed up in https://github.com/leancodepl/contractsgenerator/issues/58